File: benchmarks/zig/b10_knn.zig
// k-NN classification — Machine Learning benchmark (MICAI). Shared LCG (seed 42),
// squared Euclidean distance, vote ties to lowest class. Checksum = sum of labels.
const std = @import("std");
const M = 50000;
const Q = 10000;
const D = 8;
const K = 15;
const C = 3;
var state: u64 = 42;
fn nextDouble() f64 {
    state = state *% 6364136223846793005 +% 1442695040888963407;
    return @as(f64, @floatFromInt(state >> 33)) / @as(f64, @floatFromInt(@as(u64, 1) << 31));
}
var train: [M * D]f64 = undefined;
var label: [M]i64 = undefined;
var query: [Q * D]f64 = undefined;
pub fn main(init: std.process.Init) !void {
    const io = init.io;
    for (0..M) |t| {
        for (0..D) |d| {
            train[t * D + d] = nextDouble();
        }
        label[t] = @intFromFloat(nextDouble() * C);
    }
    for (0..Q) |q| for (0..D) |d| {
        query[q * D + d] = nextDouble();
    };
    var checksum: i64 = 0;
    var best_d: [K]f64 = undefined;
    var best_l: [K]i64 = undefined;
    for (0..Q) |q| {
        for (0..K) |j| {
            best_d[j] = 1e300;
            best_l[j] = 0;
        }
        for (0..M) |t| {
            var dist: f64 = 0.0;
            for (0..D) |d| {
                const diff = query[q * D + d] - train[t * D + d];
                dist += diff * diff;
            }
            if (dist < best_d[K - 1]) {
                var p: usize = K - 1;
                while (p > 0 and dist < best_d[p - 1]) : (p -= 1) {
                    best_d[p] = best_d[p - 1];
                    best_l[p] = best_l[p - 1];
                }
                best_d[p] = dist;
                best_l[p] = label[t];
            }
        }
        var votes: [C]i64 = .{0} ** C;
        for (0..K) |j| {
            votes[@as(usize, @intCast(best_l[j]))] += 1;
        }
        var pred: usize = 0;
        for (1..C) |c| {
            if (votes[c] > votes[pred]) pred = c;
        }
        checksum += @as(i64, @intCast(pred));
    }
    var stdout_buffer: [256]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    try stdout.print("{d}\n", .{checksum});
    try stdout.flush();
}

File: benchmarks/swift/b10_knn.swift
// k-NN classification — Machine Learning benchmark (MICAI). Shared LCG (seed 42),
// squared Euclidean distance, vote ties to lowest class. Checksum = sum of labels.
import Foundation
let M = 50000, Q = 10000, D = 8, K = 15, C = 3
var state: UInt64 = 42
@inline(__always) func nextDouble() -> Double {
    state = state &* 6364136223846793005 &+ 1442695040888963407
    return Double(state >> 33) / Double(UInt64(1) << 31)
}
var train = [Double](repeating: 0, count: M * D)
var label = [Int](repeating: 0, count: M)
for t in 0..<M { for d in 0..<D { train[t*D+d] = nextDouble() }; label[t] = Int(nextDouble() * Double(C)) }
var query = [Double](repeating: 0, count: Q * D)
for q in 0..<Q { for d in 0..<D { query[q*D+d] = nextDouble() } }
var checksum = 0
var bestD = [Double](repeating: 0, count: K)
var bestL = [Int](repeating: 0, count: K)
for q in 0..<Q {
    for j in 0..<K { bestD[j] = 1e300; bestL[j] = 0 }
    for t in 0..<M {
        var dist = 0.0
        for d in 0..<D { let diff = query[q*D+d] - train[t*D+d]; dist += diff * diff }
        if dist < bestD[K-1] {
            var p = K - 1
            while p > 0 && dist < bestD[p-1] { bestD[p] = bestD[p-1]; bestL[p] = bestL[p-1]; p -= 1 }
            bestD[p] = dist; bestL[p] = label[t]
        }
    }
    var votes = [Int](repeating: 0, count: C)
    for j in 0..<K { votes[bestL[j]] += 1 }
    var pred = 0
    for c in 1..<C where votes[c] > votes[pred] { pred = c }
    checksum += pred
}
print(checksum)
